### PR TITLE
Port isTargetAllowed to share 2.0

### PR DESCRIPTION
**Pull request by LukasReschke**
From https://github.com/owncloud/core/pull/26615

### CODE REVIEW

**Comment by LukasReschke** [comment]
Tests are failing. 🚀

File: lib/private/Files/Mount/MoveableMount.php [hunk 1]
```diff
@@ -42,4 +42,15 @@ public function moveMount($target);
...
 	public function removeMount();
+
+	/**
+	 * Returns whether this mount point is allowed to be moved into the given absolute target
+	 *
+	 * @param string $target absolute target path
+	 *
+	 * @return bool true if allowed, false otherwise
+	 *
+	 * @since 11.0
```
Comment (schiessle) [inline lib/private/Files/Mount/MoveableMount.php]: since 12.0 in our case

**Comment by rullzer** [comment]
@icewind1991 I'd like to get your opinion on this as well

File: apps/files_sharing/lib/SharedMount.php [hunk 2]
```diff
@@ -180,12 +181,59 @@ protected function stripUserFilesPath($path) {
...
+		$targetNodes = \OC::$server->getRootFolder()->getById($fileId);
+		if (empty($targetNodes)) {
+			return false;
+		}
+
+		$shareManager = \OC::$server->getShareManager();
+		$targetNode = $targetNodes[0];
+		// FIXME: make it stop earlier in '/$userId/files'
+		while (!is_null($targetNode) && $targetNode->getPath() !== '/') { 
+			$shares = $shareManager->getSharesByPath($targetNode);
```
Comment (icewind1991) [inline apps/files_sharing/lib/SharedMount.php]: Why does `byPath` want a `Node` instead of a path :see_no_evil:

**Review by icewind1991 — CHANGES_REQUESTED** [review]
I'm not happy with the way this checks if the target is shared (do we really not have an existing method to do that?), getting the node and iterating over parents is needlessly expensive

File: apps/files_sharing/lib/SharedMount.php (hunk 2 above)
Comment (rullzer) [inline apps/files_sharing/lib/SharedMount.php]: We use nodes everywhere. Was just naming stuff to be somewhat siilar to the old one (you should have complained when the code was reviewed :wink: )

**Comment by rullzer** [comment]
@icewind1991 well give me my closure table and I can optimize it properly....

**Comment by rullzer** [comment]
@icewind1991 could you have a look at the failing tests so we can get this in? I know you have some objections but lets fix them separatly.

**Comment by LukasReschke** [comment]
@icewind1991 🏓

**Comment by MorrisJobke** [comment]
Let me rebase this and get #3944 into this.

File: apps/files_sharing/lib/SharedMount.php [hunk 3]
```diff
@@ -180,12 +181,59 @@ protected function stripUserFilesPath($path) {
...
+		list($targetStorage, $targetInternalPath) = \OC\Files\Filesystem::resolvePath($target);
+		// note: cannot use the view because the target is already locked
+		$fileId = (int)$targetStorage->getCache()->getId($targetInternalPath);
+		if ($fileId === -1) {
+			// target might not exist, need to check parent instead
+			$fileId = (int)$targetStorage->getCache()->getId(dirname($targetInternalPath));
+		}
+
+		$targetNodes = \OC::$server->getRootFolder()->getById($fileId);
+		if (empty($targetNodes)) {
```
Comment (MorrisJobke) [inline apps/files_sharing/lib/SharedMount.php]: The first unit test fails here.

```
1) OCA\Files_Sharing\Tests\UpdaterTest::testDeleteParentFolder
Failed asserting that true is false.
/drone/src/github.com/nextcloud/server/apps/files_sharing/tests/UpdaterTest.php:105
```

**Comment by MorrisJobke** [comment]
@icewind1991 Please have a look at the failing tests. Would be nice to get this in and I don't properly understand this :/

**Comment by MorrisJobke** [comment]
Nothing for 12 anymore

**Pull request closed without merging**